### DISCUSSION TITLE
Effort to get Touchscreen (and models like Pro and Plus) working

### DISCUSSION THREAD

**Discussion by TheWebMachine** [mriscoc/Ender3V2S1#593] [Ideas]
Hi all! I've been watching this repo since I got my E3V2S1 Plus (and the 5W laser) a few weeks ago and came to quickly find that Creality really F***ed this firmware up! 

- Their 4x4 CR-MBL (yes, only 4x4 on a 300x300mm bed!) has some sort of algorithmic snafu that skews the z-offset in increasing amounts as it moves further up Y-axis, so the first later fails by about 2/3 back the bed
- z-offset adjustment is in 0.05mm increments! (dial-control models...and every other damn printer out there...uses 0.01mm adjustment)
- switching between FDM and Laser modes completely loses all MBL and z-offset data on the FDM side for some stupid reason, so even if you *DO* manage to get things leveled up manually in FDM, you are constantly doing it over and over again if you switch modes
- even Laser has a bug where it doesn't let you take advantage of the full 300x300mm bed and cuts off engravings at 240x240mm!

All in all, Creality really messed this up. Trying to work with their CS has been pitiful, as they clearly have no technical know-how and won't put me in touch with an actual engineer to sort through these bugs with. I also can't seem to get them to send me a dial-control (DWIN I think you call it) screen so I can confirm once and for all that it is the half-baked screen firmware that is to blame. I have a friend with a regular E3V2S1 and it doesn't have any of the issues I do, despite being nearly identical spare the screen and bed size. I really do not want to return this printer. The hardware is rock solid, 300mm3 is awesome, the laser works great, extruder is kickass...all that is missing is great firmware.

So, since I am an engineer (both electrical and software), I figured perhaps I can help give back to the printing community on this one. I noticed that the primary maintainer(s) of this repo do not have access to a Plus model and/or the touchscreen to be able to properly expand this firmware to support them, so I would like to see what I could do to help move this along for all of us. 

Is there a discord or other discussion going on right now around the touchscreen models? Any work thus far I can browse through to see where I can contribute? As it stands right now, I can only get about half my value out of this printer. I can either return it, let it sit and collect dust until Creality releases workable firmware for it, or I can keep tinkering with it and contribute my findings here so all of us who bought these particular machines can move on to better firmware.

**Comment by mriscoc** [comment]
Hi, some Creality touch screens units are closed source so the development is a little difficult for those displays. I remember that someone by mistake was able to update his touch screen firmware using the DWIN or DACAI icon set, that results in a semi-functional display, only using half of the screen and no user interaction due to lack of an encoder. So, it could be possible to work with that display but large re-work is required.

https://www.facebook.com/groups/ender3s1printer/permalink/6106210349407544/

![TouchScreen](https://user-images.githubusercontent.com/2745567/210271243-659bf657-b548-4067-90db-b9860e7dfd37.jpg)

**Comment by TheWebMachine** [comment]
Well, that looks like the touchscreen I have, so maybe I can help.

Does anyone know where I can buy a DWIN or DACAI screen (I suppose whichever one comes on the V2, ideally) that *isn't* AliExpress? lol I have the worst luck with AliExpress and avoid it whenever I can. Creality doesn't appear to sell them anymore because they are pushing everyone to upgrade to the touchscreen (with, I'm assuming, closed source firmware). I've done some looking around but haven't had much luck finding another source for them.

I figure having the other screen handy will not only allow me to get some actual use out of my printer, but also to be able to side-by-side the two screens while working on the firmware.

**Reply by lagnat** [reply]
The s1 pro comes with a touch screen afaik. Would the necessary code be here https://github.com/CrealityOfficial/Ender-3S1?  I poked around but I don't know what I'm looking for.

**Reply by rk-exxec** [reply]
I looked a bit at the code, and it seems the most difference is in the `Marlin/src/lcd`  folder in the s1_pro branch. But I have no experience with Marlin either. I had hoped you'd just need to plop the driver for the display and thats it.

**Reply by TheWebMachine** [reply]
Unfortunately, it's a bit more nuanced than that. The screen is effectively an old school "dumb terminal" that only contains assets (fonts, images, etc) and a communication protocol to talk to the board.

As with any dumb terminal, the screen and board have to be in *predetermined* agreement on what assets are available on the screen and what commands, parameters, features are present on the communication stack they are using to talk to each other. This is why you almost always have to initially re-flash the screen to switch to this firmware from stock...to ensure the board and screen are speaking exactly the same language in exactly the same syntax for everything shown or interacted with.

In turn, the touchscreen firmware from Creality needs modified to match the communication this board firmware uses and that it has all the assets this firmware might request it to show. The board firmware also likely needs to be able to identify it is talking to this type of screen to adjust layouts, communication, and expectations accordingly. Given how different an interface it is, it's going to require a lot of work on the screen, board, or both to get it 100% functional. 

It's on my projects list but, admittedly, it's a ways down that list at the moment. My work will also be limited to F4 boards if it turns out heavy modification to board firmware is required, as that is the only model (Plus) I have handy.

**Reply by rk-exxec** [reply]
I had (foolishly) assumed, that at least the functionality for which the assets are already on the screen would be easy to implement.
The only two things I really need are live z-offset and manual bed leveling, maybe nozzle preheat for filament change. For everything else I could use octoprint. If you ever decide to actually tackle this, I would be happy to help out.

**Comment by TheWebMachine** [comment]
Well, I managed to find the screen (edit: DACAI, not DWIN) that comes with the regular E3S1V2 on Amazon, in case anyone else is looking for them:

https://www.amazon.com/gp/product/B098WFCQYW

**Reply by jordathan** [reply]
someone did get it figured out somehow. i updated the firmware and the screen still works and it has ubl. i can post a pic after my daughter goes to sleep

**Reply by rk-exxec** [reply]
@jordathan Can you send the link?

**Reply by TheWebMachine** [reply]
If this fixed display firmware doesn't add 0.01mm adjustment to the z-offset, it's of little use to me (and you). My biggest complaint about Creality's touchscreen firmware was that the z-offset adjust only offered 0.05mm movements...a catastrophe for dialing in first later height and something they should have known better not to do.

The touchscreen firmware was half-baked from the start and won't be as easily updatable alongside the printer's firmware, sadly. The touchscreen would likely need a complete firmware update with each variant/version that presents different menu options and features. I don't know that the touchscreen will allow some of the same functions to be presented properly. There are lots of design considerations when going from a cursor-based interface to touch-based; whether it's not being able to fit as many finger-friendly options on a screen or the abject slowness of repeatedly touching a button to adjust an axis versus spinning a wheel. In fact, failure to take those design matters into consideration is exactly why I quickly grew to hate Creality's touchscreen.

I've been quite content having gone back to the tried-and-true dial control, despite the expense of buying the additional screen. You could probably find a used DWIN or DACAI screen for much cheaper than new. I've even started seeing Enders show up at Pawn shops/Good Will/etc...some priced way below what they're really worth since some folks just have no clue about 3D printing and still see them as little more than a toy. You could probably pick up a whole "broken" E3 worth of spare parts, including screen, for the cost of a new screen.

**Reply by rk-exxec** [reply]
I've been looking on ebay and such and found some screens for ~25€, maybe I'll just get one and be done with it. Next time I will be wiser and not trust All3DP for a buying guide for all the headaches this thing has caused me.

**Reply by TheWebMachine** [reply]
I think the general theme with Creality has always been "almost there." If you are willing to slap a few customizations on it, tweak a few things, and replace the firmware, you end up with a great printer. I have been using it more than even my Prusa lately because the direct extruder has been much kinder to some of my specialty filaments than the Prusa's bowden.

**Comment by tiagofreire-pt** [comment]
The source code for the touch screen on Ender 3 S1 Plus is now available: https://github.com/CrealityOfficial/Ender-3S1/tree/3S1_Plus_Screen

@mriscoc , would you consider add this to your project?

**Reply by mriscoc** [reply]
Sorry but that hardware is too different from the hardware that currently I have and development for it could be very time consuming.

**Reply by TheWebMachine** [reply]
I do have the touchscreen, printer, and the relevant skillset...but I am juggling a lot of pet projects at the moment and haven't setup a build environment for this yet. It will likely be 2ish months before I'll get the chance to look into this in earnest.
